### ISSUE TITLE
feat: Trial generation beyond make_trials, for sequential experimentation

### ISSUE THREAD

**Issue by sunkusun9** [sunkusun9/ml-labs#132]
## Purpose

Support experiments that proceed in rounds, where the next Trial follows from the previous result —
Bayesian parameter optimization, feature selection, early-stopping calibration. Today the library
can run a batch of Trials and record their outcomes, but has no facility for *producing* the next
batch, so every round is hand-assembled.

## Background

`make_trials` sweeps a `param_grid` over one fixed processor. Everything else — processor, adapter,
edges, method — is constant across the sweep:

```python
make_trials('lgbm', processor='lightgbm.LGBMClassifier',
            edges={'X': 'scaler:(*)', 'y': '{target}'},
            param_grid={'max_depth': [3, 5], 'learning_rate': [0.05, 0.1]})
```

That is a one-dimensional sweep along params. It is not the axis real rounds move along.

`examples/kaggle_s6e6/2. Second Phase.ipynb` runs five rounds, and what changes between them is
**edges** — each round adds a feature block — while **model family** varies within a round, each
family needing its own edges. Neither is expressible, so the notebook builds four helpers:

```python
MODELS = {'xgb': dict(processor=..., adapter=..., params=...), 'lgb': ..., 'cb': ..., 'nn': ..., 'lr': ...}

def trial(name, model, X, **params):
    m = MODELS[model]
    return Trial(name, m['processor'], {'X': X, **y_edges}, method='predict',
                 adapter=m.get('adapter'), params={**m['params'], **params})

def edges_for(*extra, xgb_cols=None):
    ns = ''.join(f' + {n}:(*)' for n in extra)
    return {'xgb': dsl_set(...) + ' + coov:(*)' + ns, 'lgb': ..., 'cb': ..., 'nn': ..., 'lr': ...}

def round_trials(idx, *extra, **kw):
    E = edges_for(*extra, **kw)
    return [trial(f'{m}{idx}', m, E[m]) for m in ('xgb', 'lgb', 'cb', 'nn', 'lr')]
```

Three things are being reinvented here: a **catalog** of model families (processor + adapter +
default params under a short key), a sweep whose axis is **edges**, and a **product** of family ×
round with a naming scheme.

The sequential part appears too, in its crudest form — cell 45 derives new Trials from Trials that
already ran, using a number read out of their own collected eval curves:

```python
def stack_trial(name, src, model, n_rounds):
    params = {k: v for k, v in src.params.items() if k not in NO_ES}
    params['epochs' if model == 'nn' else 'n_estimators'] = n_rounds
    return Trial(name, src.processor, dict(src.edges), method='predict_proba',
                 adapter=src.adapter, params=params)
```

The inputs for this already exist and are queryable: `TrialStore` holds definitions and per-fold
history, `MetricCollector` holds scores, `ModelAttrCollector` holds feature importances and eval
curves. `Trial` is plain data, so producing them programmatically is natural. What is missing is
anything between "read the results" and "here is the next batch".

## Direction

Two layers, and they should be judged separately.

**1. Trial construction.** Something more granular than `make_trials` — able to sweep along edges,
to carry a reusable model catalog, and to produce a named product across axes. Deterministic naming
matters as much as the generation: names are identity in `TrialStore`, so a regenerated round must
produce the same names or it silently forks the history.

**2. Sequential rounds.** A shape for "given what ran, produce the next batch" — the step a
Bayesian optimizer or a feature-selection loop plugs into. The open question is how much of this
belongs in the library at all versus being a loop the caller writes with an optimizer of their
choice; the library's job may end at making the read-and-generate cycle expressible.

## Open questions

1. Is the model catalog a first-class object, or is a plain dict passed to a generator function
   enough? A dict is already what the notebook uses and it works; making it a type buys validation
   and a place to hang defaults.
2. What are the sweep axes? params (exists), edges, processor/family — and are they a product, or
   zipped, or explicitly enumerated per combination?
3. Naming. `make_trials` uses `{name}_{idx}` zero-padded from a sorted grid. A multi-axis product
   needs a scheme that stays stable when an axis gains a value — otherwise adding one model family
   renames every later trial and orphans its history.
4. Does layer 2 need anything persisted? A round is currently implicit — nothing records that
   `xgb2`/`lgb2`/`cb2` were one round, or that round 3 followed from round 2's result. `Trial.tag`
   exists and is unused for this.
5. Relationship to #131: a generated batch is a natural thing to emit as a document before running
   it, which is exactly the "propose, then run" shape. Should generation produce Trials, or a
   document that materializes into Trials?
6. Where does `stack_evals_result` fit? The best-iteration extraction the notebook writes by hand
   (`stop_rounds`, cell 42) is the same read for every gradient-boosting family, and #63 flagged it
   before the redesign.

**Comment by sunkusun9** [comment]
## Carried over from #63

Two items from #63 that outlived the redesign and were not otherwise absorbed.

**HP-tuning visualisation.** A parameter-vs-score scatter is written by hand after every sweep.
This sits directly on layer 1 above: once a sweep knows its own axes, the axis values are what a
plot needs, and they are currently only recoverable by parsing them back out of trial names or
re-reading `TrialStore.trials`. Whether the library plots anything is a separate question from
whether a sweep result carries its axes.

**Trainer chaining.** Stacking inference nests one Trainer's output into the next by hand:

```python
next(trainer1.process(...)).data   # fed into trainer2.process()
```

Weaker fit here — this is inference-time composition, not Trial generation. Recorded so it is not
lost; if it turns out to belong on its own, split it back out.

**Comment by sunkusun9** [comment]
## Naming: settled

**A Trial name carries no content.** Encoding the configuration into the name holds for two or
three axes and then stops — by the fifth round of a real sweep, the name is neither readable nor
stable, and reading it is no longer how anyone finds out what ran.

Every mature system landed here, from opposite directions:

- **Experiment trackers assign meaningless handles.** MLflow generates `adjustable-hound-114` next
  to a UUID run id, W&B `sunny-dawn-3`, Optuna and Sacred use a plain sequence number within a
  study. The configuration lives in queryable columns, and the compare view *computes* which of
  them differ across the selected runs.
- **Lab notebooks have used `ABC-2024-017` for centuries** — initials, year, sequence. Chemistry
  registers compounds as `JS-1-42`, clinical trials as `NCT01234567`, sequences as accession
  numbers. The reason is not convenience: a name must be assignable *before* the result is known,
  and is never reused, not even for a failed run.
- **The counterexample is instructive.** CMS/ATLAS dataset names *are* long and descriptive
  (`/QCD_Pt-15to7000_TuneCP5_13TeV/RunIISummer20UL18MiniAODv2-106X.../MINIAODSIM`), because
  thousands of people share them with no metadata service at hand. A descriptive name is what you
  write when the axes are **not** queryable.

Two rules come out of that:

1. **A name may be a handle when the axes are queryable.** `TrialStore.trials` already holds
   processor / method / adapter / params / edges / tag as columns, so they are. A descriptive name
   would be a second, worse copy of what the store already answers.
2. **Assign, never derive.** A sequence number survives an axis gaining a value because it was
   never computed from the axes in the first place. This is the property, and it is exactly what
   `make_trials` lacks.

### `make_trials` derives its names, and is already broken by it

```python
width = len(str(len(combos) - 1)) if len(combos) > 1 else 0
name if len(combos) == 1 else f"{name}_{idx:0{width}d}"
```

| change | result |
|---|---|
| 1 combo → 2 | `lgbm` → `lgbm_0`, `lgbm_1` — the original name is gone |
| 10 combos → 11 | `lgbm_0` → `lgbm_00` — **every name changes** (`test_wide_index_padding` pins this) |
| `max_depth: [3,5]` → `[3,5,7]` | `lgbm_2` moves from `(lr .1, md 3)` to `(lr .05, md 7)` — same name, different configuration |

The first two are **silent**. A rename produces a *new* name, so `Project.set_trial`'s freeze gate
never fires — that name has no history to protect. The old name keeps its definition and its
results, the new one runs from scratch, and nothing says they were the same sweep. Only the third
raises, and it reads to the caller as "this axis cannot gain a value".

So this is a defect in the function that exists today, not a shortcoming of its single-axis scope.

## Reading: the device exists, on the wrong side of the boundary

If names say nothing, something else has to make a Trial's content legible. That something is
already written — `_describer.compare_nodes`:

- groups by processor
- expands every param, then keeps **only the columns that differ**
  (`diff_cols = [c for c in df.columns if len({repr(v) for v in df[c]}) > 1]`)
- splits `edges['X']` into source-node segments, folds the shared variables into the column header
  (`DataSource [a, b, c]`) and leaves only the per-node difference in the cell

That is the same thing W&B's compare view does, and it was written when models were `role='head'`
nodes. Models left for `Trial` in #123 and this stayed behind on the builder. The traces are
visible: `docs/guide/project-pipeline.md:232` still writes `p.compare_nodes(['m1', 'm2'])` — model
names that can no longer be nodes — and the s6e6 notebook never calls it, because what it wants to
compare is a Trial.

**Trials lost a reader when they left the pipeline.**

Moving it costs almost nothing. `compare_nodes` only ever touches `spec_map[n].processor`,
`.params` and `.edges` — it already runs on `ProcessorSpec` and is coupled to the pipeline solely
at its entrance (`spec_map = {n: pipeline.get_node_spec(n) for n in nodes}`). `ProcessorSpec`
exists precisely so node / Trial / Predictor are one thing. Extract `compare_specs({name: spec})`,
have `compare_nodes` call it, and the Trial side is `[t.get_spec() for t in trials]`.

### This dissolves open question 2

The issue asks what the sweep axes are — params, edges, family; product, zipped or enumerated.
**If the reader computes what differs, the axes never have to be declared at authoring time.** They
are discovered from the store, whether the Trials came from a generator or a hand-written loop.

Which reorders this issue: the catalog and the sweep axes are conveniences, and the reader is the
thing that is actually load-bearing. A generator that is hard to read is worse than a loop that is
easy to read.

## Lineage

Two things the diff table cannot answer, both of which the notebook needs and neither of which is
recorded anywhere today:

1. **Grouping** — that `xgb5` / `lgb5` / `cb5` / `nn5` / `lr5` were one round. The `5` in the name
   is currently the only record, which is precisely what "names carry no content" removes.
2. **Derivation** — cell 62's `stack_trial(f'{name}_stk', ALL_TRIALS[name], m, n)` builds a new
   Trial from one that already ran, fixing `n_estimators` to a best-iteration read out of that
   Trial's own collected eval curve. `src` is a parameter of the function and evaporates at the
   store boundary. Without it, that number has no explanation.

These are **two mechanisms, not one**, and W3C PROV splits them at the same seam: `wasGeneratedBy`
(an Activity produced these) versus `wasDerivedFrom` (this came from that Entity). Collapsing them
makes one of the two lie — a round tag cannot say that round 5 followed from round 4's result, and
a parent pointer cannot say that five Trials ran together.

Worth knowing how rare the second one is. **Config lineage is largely absent from ML tooling**:
MLflow's `parentRunId` is containment (sweep → trial), not derivation; W&B's lineage graph is over
*artifacts*, not configurations; Optuna records no parent at all. The reason is that in an ordinary
sweep, trials are siblings from a generator — the parent is the sweep, and a parent pointer would
carry no information.

Derivation carries information only in **adaptive** experimentation, and there it is essential:
successive halving promotes a configuration to a larger budget, PBT clones weights from another
trial and perturbs its hyperparameters, a warm-started optimizer continues a prior study. The
sciences that manage by lineage are the ones with the same shape — chemistry registers a compound
together with its synthetic route, cell lines are worthless without passage history.

`stack_trial` is exactly this case, which is why the concept fits here at all.

### The precedent is already in the codebase

`Predictor.src_trial` / `src_experimenter` *is* `wasDerivedFrom`, and `_predictor_store.py:58`
shows the policy already chosen:

```sql
src_trial        TEXT,
src_experimenter TEXT,
```

No foreign key, no validation. `Project.remove_trial` does not touch Predictors, so the field can
outlive what it names — and that is correct. Provenance records *what was the case then*, not a
reference that must still resolve. A discarded sample number is not erased from a lab notebook.

So a Trial→Trial link would be the second application of a settled policy, not a new one.

### The condition that makes it survive

Lineage is cheap to write and expensive to keep honest. `Predictor.src_trial` works because
`from_trial()` is the **single** promotion path and fills it mechanically. Asked to write
`Trial(name, ..., src_trial='lgb5')` by hand, nobody will, and the column ends up mostly NULL —
present but not trustworthy, which is worse than absent.

So the field only pays for itself alongside a derivation path that fills it — the role
`from_trial` plays for Predictor, and the place `stack_trial` already occupies in the notebook.

## MLflow was considered and set aside

Adopting an external tracker was weighed. It does not fit, for reasons that are structural rather
than aesthetic:

- **The unit differs.** MLflow's unit is a run — one configuration, trained once. Here it is Trial
  × `(outer_idx, inner_idx)`; one name fans out across a fold grid. `MetricCollector`'s key is
  `(node, idx, inner_idx, split)`, four dimensions against MLflow metrics' `(key, step)`. Either a
  fold becomes a run and a Trial becomes a nested group, or a Trial becomes a run and the fold grid
  gets encoded into `step`. Both lose a dimension.
- **Params are not flat here.** They carry ref specs
  (`{'__ref__': 'mllabs.ColSelector', '__params__': {...}}`) and DSL edge strings. Flattened to
  string key-values, the diff view above would be comparing opaque strings — losing the exact thing
  it exists to provide.
- **The Pipeline is shared, versioned state**, not per-run config. There is nowhere in a run to put
  "the preprocessing graph, v5" except duplicated into every run or uploaded as an opaque artifact.
- **MLflow records; it does not re-execute.** Staleness, `'built'` fold skipping and reset cascades
  stay here regardless, so the outcome is two stores that can disagree.

What it would actually buy is a UI, and that does not advance milestone #10 — an agent needs state
readable as data (#130) and definitions writable as documents (#131), neither of which a run table
provides, while flattened params move #130 backwards.

Export stays open as a later option, and `Collector` is the seam for it — results already leave
through there, and optional dependencies are established practice. But **an export can only carry
what the native model distinguishes**, so it is worth nothing until spec-diff, grouping and lineage
exist here first.

## Where this leaves the issue

Settled:

- Trial names carry no content; identity is a handle, assigned rather than derived.
- The reader, not the generator, is the load-bearing half of layer 1.
- Grouping and derivation are two mechanisms.
- No external tracker as the store.

Open:

- Which naming scheme replaces the derived index in `make_trials`, and what happens to the existing
  `{name}_{idx}` behaviour.
- Whether grouping is `Trial.tag` (which exists and is unused for this) or something with a place
  of its own.
- Whether the Trial→Trial link is worth a slot, given it only survives with a derivation path
  alongside it.
- Whether layer 2 (sequential rounds) should split out into its own issue — it depends on #131 for
  whether generation emits Trials or a document, while layer 1 does not.
